### PR TITLE
DTensor: add comm tests to test_tp_examples

### LINKED ISSUE #121649: guard on tensor parallelism test examples
This task is to add some guards to the tensor parallelism test examples to ensure we incurred the right collectives: https://github.com/pytorch/pytorch/blob/main/test/distributed/tensor/parallel/test_tp_examples.py

Specifically we can follow this logic to assert on the communication collectives incurred using the CommDebugMode https://github.com/pytorch/pytorch/blob/main/test/distributed/tensor/parallel/test_tp_examples.py#L100

The test example we want to add this guard on is the Transformer e2e training example https://github.com/pytorch/pytorch/blob/main/test/distributed/tensor/parallel/test_tp_examples.py#L179

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

### PR DESCRIPTION
**Pull request by d4l3k**
This adds some basic comm tests to test_tp_examples. This validates that the expected distributed calls are being made for `test_transformer_training`.

Fixes #121649

Test plan:

```
pytest test/distributed/tensor/parallel/test_tp_examples.py -k test_transformer_training
```

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

### CODE REVIEW

**Comment by awgu** [comment]
Oof, this will conflict with https://github.com/pytorch/pytorch/pull/121660 -- we want to have a canonical TP/SP sharding for the `Transformer` so we can use that for FSDP + TP/SP unit tests.
cc: @wanchaol 

Should we just duplicate the TP/SP sharding logic since it seems we need to insert inline tests here?

**Comment by d4l3k** [comment]
@awgu I can just pull the tests to wrap the whole parallelize method since that shouldn't be making any network calls. Not a problem. Do you know when that's landing?

File: test/distributed/tensor/parallel/test_tp_examples.py [hunk 1]
```diff
@@ -209,11 +209,14 @@ def test_transformer_training(self, is_seq_parallel=False):
                 "pos_embeddings": ColwiseParallel(output_layouts=Replicate()),
             }
 
-        model_tp = parallelize_module(
-            model_tp,
-            device_mesh,
-            root_plan
-        )
+        with CommDebugMode() as comm_mode:
+            model_tp = parallelize_module(
```
Comment (wanchaol) [inline test/distributed/tensor/parallel/test_tp_examples.py:213]: Ohh I think we want to track the collectives in runtime forward/backward, we don't need to track the sharding part (parallelize_module)

File: test/distributed/tensor/parallel/test_tp_examples.py [hunk 2]
```diff
@@ -239,7 +242,10 @@ def test_transformer_training(self, is_seq_parallel=False):
                 output_layouts=Shard(1)
             ) if is_seq_parallel else RowwiseParallel()
 
-            parallelize_module(layer, device_mesh, layer_parallelize_plan)
+
+            with CommDebugMode() as comm_mode:
+                parallelize_module(layer, device_mesh, layer_parallelize_plan)
+            self.assertDictEqual(comm_mode.get_comm_counts(), {})
```
Comment (wanchaol) [inline test/distributed/tensor/parallel/test_tp_examples.py:248]: ditto: no need to track collectives on the parallelize_module parts

File: test/distributed/tensor/parallel/test_tp_examples.py [hunk 3]
```diff
@@ -255,7 +261,10 @@ def test_transformer_training(self, is_seq_parallel=False):
                 input_layouts=Shard(1),
                 output_layouts=Replicate(),
             ) if is_seq_parallel else RowwiseParallel(input_layouts=Replicate())
-        parallelize_module(model_tp.output, device_mesh, output_parallelize_plan)
+
+        with CommDebugMode() as comm_mode:
+            parallelize_module(model_tp.output, device_mesh, output_parallelize_plan)
```
Comment (wanchaol) [inline test/distributed/tensor/parallel/test_tp_examples.py:266]: ditto

File: test/distributed/tensor/parallel/test_tp_examples.py [hunk 4]
```diff
@@ -282,18 +291,48 @@ def test_transformer_training(self, is_seq_parallel=False):
...
 
         # Ensure model weights are still the same after update.
         optim.step()
-        optim_tp.step()
+        with CommDebugMode() as comm_mode:
+            optim_tp.step()
         self._check_module(model, model_tp)
+        if is_seq_parallel:
+            self.assertDictEqual(comm_mode.get_comm_counts(), {
+                c10d_functional.all_reduce: 30,
```
Comment (wanchaol) [inline test/distributed/tensor/parallel/test_tp_examples.py:332]: wat, I am shocked that there're 30 allreduces in optimizer step 😮 I was expecting around 5 allreduces hmm

could you pass `foreach=True` to Adam optimizer and see if that would reduce # allreduces?

**Review by wanchaol — COMMENTED** [review]
Nice and fast changes! have some comments around the sharding init comm tracking

File: test/distributed/tensor/parallel/test_tp_examples.py (hunk 4 above)
Comment (yifuwang) [reply test/distributed/tensor/parallel/test_tp_examples.py:332]: I think foreach is a single kernel but the tensors are still kept separate

Comment (tianyu-l) [reply test/distributed/tensor/parallel/test_tp_examples.py:332]: It seems each LayerNorm would incur 2 all-reduces, weight & bias. With default two-layer Transformer, we have two `attention_norm`, two `ffn_norm`, and one final `norm` -- adding up to 10 all-reduces already?

In general, I wonder if we should include a breakdown of the collectives in the test -- ideally it should be a function on the number of TransformerBlocks (and other relevant configs), so even if we modify the model architecture, the test could still pass.

Comment (d4l3k) [reply test/distributed/tensor/parallel/test_tp_examples.py:251]:
```
Traceback (most recent call last):
  File "/home/tristanr/pytorch/torch/testing/_internal/common_distributed.py", line 540, in wrapper
    self._join_processes(fn)
  File "/home/tristanr/pytorch/torch/testing/_internal/common_distributed.py", line 759, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/home/tristanr/pytorch/torch/testing/_internal/common_distributed.py", line 809, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Process 0 exited with error code 10 and exception:
Traceback (most recent call last):
  File "/home/tristanr/pytorch/torch/testing/_internal/common_distributed.py", line 656, in run_test
    getattr(self, test_name)()
  File "/home/tristanr/pytorch/torch/testing/_internal/common_distributed.py", line 542, in wrapper
    fn()
  File "/home/tristanr/pytorch/torch/testing/_internal/common_utils.py", line 2739, in wrapper
    method(*args, **kwargs)
  File "/home/tristanr/pytorch/torch/testing/_internal/common_utils.py", line 439, in instantiated_test
    test(self, **param_kwargs)
  File "/home/tristanr/pytorch/torch/testing/_internal/distributed/_tensor/common_dtensor.py", line 374, in wrapper
    func(self, *args, **kwargs)  # type: ignore[misc]
  File "/home/tristanr/pytorch/torch/testing/_internal/common_distributed.py", line 181, in wrapper
    return func(*args, **kwargs)
  File "/home/tristanr/pytorch/test/distributed/tensor/parallel/test_tp_examples.py", line 247, in test_transformer_training
    optim_tp.step()
  File "/home/tristanr/pytorch/torch/optim/optimizer.py", line 391, in wrapper
    out = func(*args, **kwargs)
  File "/home/tristanr/pytorch/torch/optim/optimizer.py", line 76, in _use_grad
    ret = func(self, *args, **kwargs)
  File "/home/tristanr/pytorch/torch/optim/adam.py", line 168, in step
    adam(
  File "/home/tristanr/pytorch/torch/optim/adam.py", line 318, in adam
    func(params,
  File "/home/tristanr/pytorch/torch/optim/adam.py", line 522, in _multi_tensor_adam
    torch._foreach_lerp_(device_exp_avgs, device_grads, 1 - beta1)
  File "/home/tristanr/pytorch/torch/distributed/_tensor/api.py", line 279, in __torch_dispatch__
    return DTensor._op_dispatcher.dispatch(
  File "/home/tristanr/pytorch/torch/distributed/_tensor/dispatch.py", line 111, in dispatch
    op_info = self.unwrap_to_op_info(op_call, args, kwargs)
  File "/home/tristanr/pytorch/torch/distributed/_tensor/dispatch.py", line 314, in unwrap_to_op_info
    raise RuntimeError(
RuntimeError: aten._foreach_lerp_.Scalar: got mixed torch.Tensor and DTensor, need to convert all torch.Tensor to DTensor before calling distributed operators!
```

Seems to error out with `foreach=True`

File: test/distributed/tensor/parallel/test_tp_examples.py (hunk 2 above)
Comment (d4l3k) [reply test/distributed/tensor/parallel/test_tp_examples.py]: done

File: test/distributed/tensor/parallel/test_tp_examples.py (hunk 3 above)
Comment (d4l3k) [reply test/distributed/tensor/parallel/test_tp_examples.py]: done

File: test/distributed/tensor/parallel/test_tp_examples.py (hunk 4 above)
Comment (awgu) [reply test/distributed/tensor/parallel/test_tp_examples.py:251]: If you want to bypass that error:
```
from torch.distributed._tensor.experimental import implicit_replication

with implicit_replication():
    optim.step()
```
cc: @wanchaol

Comment (wanchaol) [reply test/distributed/tensor/parallel/test_tp_examples.py:251]: Ohh i see, this is nn.Layernorm not RMSNorm, where it have 2 allreduces per-layernorm, this make sense then

**Review by wanchaol — APPROVED** [review]
This lgtm! thanks for the fast PR!

**Pull request closed without merging**